sites/partners/src/listings/PaperListingForm/sections/BuildingDetails.tsx

```diff
@@ -8,13 +8,116 @@ import {
   ViewItem,
   Select,
   stateKeys,
+  FieldGroup,
+  ListingMap,
+  LatitudeLongitude,
 } from "@bloom-housing/ui-components"
+import { FormListing } from "../index"
+import GeocodeService from "@mapbox/mapbox-sdk/services/geocoding"
 
-const BuildingDetails = () => {
+interface MapBoxFeature {
+  center: number[] // Index 0: longitude, Index 1: latitude
+}
+
+interface MapboxApiResponseBody {
+  features: MapBoxFeature[]
+}
+
+interface MapboxApiResponse {
+  body: MapboxApiResponseBody
+}
+
+type BuildingDetailsProps = {
+  listing?: FormListing
+  latLong?: LatitudeLongitude
+  setLatLong?: (latLong: LatitudeLongitude) => void
+  customMapPositionChosen?: boolean
+  setCustomMapPositionChosen?: (customMapPosition: boolean) => void
+}
+
+const BuildingDetails = ({
+  listing,
+  setLatLong,
+  latLong,
+  customMapPositionChosen,
+  setCustomMapPositionChosen,
+}: BuildingDetailsProps) => {
   const formMethods = useFormContext()
 
   // eslint-disable-next-line @typescript-eslint/unbound-method
-  const { register } = formMethods
+  const { register, control, getValues } = formMethods
+
+  interface BuildingAddress {
+    city: string
+    state: string
+    street: string
+    zipCode: string
+  }
+
+  const buildingAddress: BuildingAddress = useWatch({
+    control,
+    name: "buildingAddress",
+  })
+
+  const mapPinPosition = useWatch({
+    control,
+    name: "mapPinPosition",
+  })
+
+  const displayMapPreview = () => {
+    return (
+      buildingAddress.city &&
+      buildingAddress.state &&
+      buildingAddress.street &&
+      buildingAddress.zipCode &&
+      buildingAddress.zipCode.length >= 5
+    )
+  }
+
+  let geocodingClient
+  if (process.env.mapBoxToken || process.env.MAPBOX_TOKEN) {
+    geocodingClient = GeocodeService({
+      accessToken: process.env.mapBoxToken || process.env.MAPBOX_TOKEN,
+    })
+  }
```

Comment:
Can you wrap this with a useEffect, so we only create the client once. I think you'll either have to move `let geocodingClient` above the component or use useState to set it.